core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -1005,13 +1009,24 @@ object Parser {
 
   /** Repeat this parser 0 or more times, but no more than `max`
     *
-    * @throws java.lang.IllegalArgumentException if max < 1
+    * It may seem weird to accept 0 here, but without, composing
```

Comment:
we call these functions from each other in a pretty confusing pattern. Once I was biasing the repetition count to small numbers, the tests were failing because we couldn't handle 0 here. I figured users could get confused too in complex calling situations and I figured 0 should be a valid value here.

Open to reconsidering.

Reply:
hm, this indeed seems confusing as with this `p1` just get's "ignored" right?

If I'm seeing this right one of the main issues is due to `repSep`. i.e for `repSep(string(...), 1, 1, string(...)))`

```scala
      // rest would be ignored
      val rest = (sep.void.with1.soft *> p1).rep0(min = min - 1, max = max - 1)
      (p1 ~ rest).map { case (h, t) => NonEmptyList(h, t) }
```

could we optimize it for:

```scala
    if (max == 1) {
      p1.map(NonEmptyList.one(_))
    } else {
      val rest = (sep.void.with1.soft *> p1).rep0(min = min - 1, max = max - 1)
      (p1 ~ rest).map { case (h, t) => NonEmptyList(h, t) }
    }
```

this way I guess we can keep the requirement on max >= 1. tests would need to be updated to make sure max >= 1

Reply:
If we do accept 0 here, we probably should for all `rep0` variations to be at least consistent.

Reply:
We do accept 0 everywhere that is a Parser0. We don't accept for Parser, since it has to consume something.

Reply:
@regadas we can't actually optimize it when the max == 1, because that means 0 or 1. We could use oneOf for that, but I doubt it will actually be faster.

Reply:
@johnynek hm, maybe i'm sewing it wrong but the optimization would be on `def repSep[A](p1: Parser[A], min: Int, max: Int, sep: Parser0[Any]): Parser[NonEmptyList[A]]` which repeats 1 or more. 

The main idea here is not really the optimization but it will avoid calling `rep0(0, 0)` L1090